script/generate-release-notes.ts

```diff
@@ -0,0 +1,304 @@
+
+  for (const entry of changelogForVersion) {
+    const externalMatch = releaseEntryExternalContributor.exec(entry)
+    if (externalMatch) {
+      const category = parseCategory(externalMatch[1])
+      const text = externalMatch[2].trim()
+      const ids = extractIds(externalMatch[3])
+      const contributor = externalMatch[4]
+
+      if (!category) {
+        console.warn(`unable to identify category for '${entry}'`)
+      } else {
+        releaseNotesByGroup[category].push({
+          text,
+          ids,
+          contributor,
+        })
+      }
+    } else {
+      const match = releaseEntryRegex.exec(entry)
+      if (match) {
+        const category = parseCategory(match[1])
+        const text = match[2].trim()
+        const ids = extractIds(match[3])
+        if (!category) {
+          console.warn(`unable to identify category for '${entry}'`)
+        } else {
+          releaseNotesByGroup[category].push({
+            text,
+            ids,
+          })
+        }
+      } else {
+        console.warn(`release entry does not match any format: '${entry}'`)
+      }
+    }
+  }
+
+  return releaseNotesByGroup
+}
+
+function formatReleaseNote(note: ReleaseNoteEntry): string {
+  const idsAsUrls = note.ids
+    .map(id => `https://github.com/desktop/desktop/issues/${id}`)
+    .join(' ')
+  const contributorNote = note.contributor
+    ? `. Thanks ${note.contributor}!`
+    : ''
+
+  const template = ` - ${note.text} - ${idsAsUrls}${contributorNote}`
+
+  return template.trim()
+}
+
+function renderSection(
+  name: string,
+  items: Array<ReleaseNoteEntry>,
+  omitIfEmpty: boolean = true
+): string {
+  if (items.length === 0 && omitIfEmpty) {
+    return ''
+  }
+
+  const itemsText =
+    items.length === 0 ? 'TODO' : items.map(formatReleaseNote).join('\n')
+
+  return `
+## ${name}
+
+${itemsText}
+  `
+}
+
+function formatEntry(e: ChecksumEntry): string {
+  return `${e.checksum} ${e.filename}`
+}
+
+function renderArchitectureIfNotEmpty(
+  name: string,
+  items: Array<ChecksumEntry>
+): string {
+  if (items.length === 0) {
+    return ''
+  }
+
+  const itemsText = items.map(formatEntry).join('\n')
+
+  return `
+## ${name}
+
+${itemsText}`
+}
+
+/**
+ * Takes the release notes entries and the SHA entries, then merges them into the full draft release notes ✨
+ */
+function generateDraftReleaseNotes(
+  releaseNotesGroups: ReleaseNotesGroups,
+  shaEntries: ChecksumGroups
+): string {
+  const draftReleaseNotes = `
+${renderSection('New', releaseNotesGroups.new)}
+${renderSection('Added', releaseNotesGroups.added)}
+${renderSection('Fixed', releaseNotesGroups.fixed, false)}
+${renderSection('Improved', releaseNotesGroups.improved, false)}
+${renderSection('Removed', releaseNotesGroups.removed)}
+
+## SHA-256 checksums
+${renderArchitectureIfNotEmpty('x64', shaEntries.x64)}
+${renderArchitectureIfNotEmpty('ARM64', shaEntries.arm64)}
+${renderArchitectureIfNotEmpty('ARM', shaEntries.arm)}`
```

Comment:
I'm going to try uploading the `.sha256` checksum files rather than including these inline releases. We should also clean this up.